Fix golint issues in plugin/pkg/admission

miguelbernadi opened the pull request:
**What this PR does / why we need it**:
Fix golint violations in plugin/pkg/admission tree, as reported in #68026 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
I've added some comments were guidance would be welcome regarding how to tackle some issues, or if to ignore them in this PR.

They will all be squashed before merging.
**Release note**:
```
NONE
```

miguelbernadi committed:
Fix golint violations in several plugins

miguelbernadi commented:
/kind cleanup

miguelbernadi commented on plugin/pkg/admission/eventratelimit/apis/eventratelimit/register.go:
```diff
@@ -22,8 +22,11 @@ import (
 
 var (
+	// SchemeBuilder is a pointer used to call AddToScheme
 	SchemeBuilder = runtime.NewSchemeBuilder(addKnownTypes)
```
I tried unexporting `SchemeBuilder` and tests would pass, but other packages in the tree are using the same scheme and some can't be unexported. Therefore I keep it here for consistency.

miguelbernadi commented on plugin/pkg/admission/limitranger/interfaces.go:
```diff
@@ -22,6 +22,7 @@ import (
 )
 
+// LimitRangerActions is an interface defining actions to be carried over ranges to identify and manipulate their limits
 type LimitRangerActions interface {
```
golint recommends renaming this interface to `Actions` to prevent stutter. This is the last violation remaining in the package and seems reasonable, but would break the interface for consumers. A quick search of users inside the repo shows only 2 uses:
```
vendor/github.com/kubernetes/repo-infra/kazel/kazel.go
613:	options := []string{"limitranger.LimitRangerActions", "BUILD"}

vendor/github.com/bazelbuild/bazel-gazelle/internal/config/config.go
93:var DefaultValidBuildFileNames = []string{"limitranger.LimitRangerActions", "BUILD"}
```

miguelbernadi replied:
That could work. I haven't addressed this yet in case there are other opinions. I'll tackle that once tests are green again.

miguelbernadi replied:
It's done.

liggitt replied:
> This is the last violation remaining in the package and seems reasonable, but would break the interface for consumers

renaming the package also breaks the interface

miguelbernadi commented:
golint suggests exporting the type `nodePlugin` [definition](https://github.com/kubernetes/kubernetes/pull/69941/commits/c60615f11fc7690bd57438c5d6c61464ffa395cb#diff-cbf05e22a4f412de387669881892d46bR63) as it's returned to users but still private, which may be cumbersome. It has plenty of public methods but all fields are private, so it may not be much of an issue to make it public.

It's the only violation left in the package.

miguelbernadi commented:
golint recommends renaming the struct `PodSecurityPolicyPlugin` to `Plugin` to avoid stutter. There's few usages of this name inside the repo, but it would affect all consumers.

yue9944882 commented:
/ok-to-test

yue9944882 commented:
>  it may be worth splitting the PR into individual PRs. Willing to do it if it's considered worth it.

one pull is fine. just squash after we are sure its ready to be merged.

yue9944882 commented on plugin/pkg/admission/eventratelimit/apis/eventratelimit/register.go:
```diff
@@ -22,8 +22,11 @@ import (
+	// SchemeBuilder is a pointer used to call AddToScheme
 	SchemeBuilder = runtime.NewSchemeBuilder(addKnownTypes)
-	AddToScheme   = SchemeBuilder.AddToScheme
+	// AddToScheme is used to register the types to API
```
no need for a second line

miguelbernadi replied:
I'll address all instances of it ASAP

miguelbernadi replied:
All instance are now oneline.

yue9944882 commented:
> golint suggest the same for podNodeSelector, podPreset, podTolerations and serviceAccount as for nodePlugin. Once the types are exported there will probably be some stutter complaint.

exporting these will be fine i suppose. renaming them to `Plugin`

miguelbernadi commented:
> > golint suggest the same for podNodeSelector, podPreset, podTolerations and serviceAccount as for nodePlugin. Once the types are exported there will probably be some stutter complaint.
> 
> exporting these will be fine i suppose. renaming all them to `Plugin`

I'll add that renaming and consequences in a new commit, probably tonight GMT timezone.

miguelbernadi commented:
@yue9944882 The renaming of the structures is finished and I removed several packages from the golint blacklist. There are several test failures, so I'm trying to find out where did I break them. Some seem flaky but most are directly related to the code.

miguelbernadi commented:
I missed verifying the tests for podtolerationrestriction and a test was broken. The fix is squashed in the renaming commit.

miguelbernadi commented:
/retest pull-kubernetes-e2e-kops-aws

miguelbernadi commented:
/test pull-kubernetes-e2e-kops-aws

miguelbernadi commented:
Rebased to resolve conflicts with master.

miguelbernadi commented:
/test pull-kubernetes-integration

miguelbernadi commented:
@yue9944882 @derekwaynecarr @deads2k All suggested work is complete, review comments addressed and tests are passing. Anything else to be done or should I squash everything into a single commit?

yue9944882 commented on plugin/pkg/admission/podtolerationrestriction/doc.go:
```diff
@@ -0,0 +1,30 @@
+// scheduler.alpha.kubernetes.io/defaultTolerations and
+// scheduler.alpha.kubernetes.io/tolerationsWhitelist annotations
+// keys.
+package podtolerationrestriction // import "k8s.io/kubernetes/plugin/pkg/admission/podtolerationrestriction"
```
seems that its not required by golint, but it's still a good change tho.

miguelbernadi replied:
The comment was for a function I had to document and was a good description for the package, so I moved it and adjusted both to golint styles.

yue9944882 commented on hack/.golint_failures:
```diff
@@ -422,23 +422,9 @@ pkg/volume/util/fs
 pkg/volume/vsphere_volume
-plugin/pkg/admission/antiaffinity
-plugin/pkg/admission/eventratelimit/apis/eventratelimit
 plugin/pkg/admission/eventratelimit/apis/eventratelimit/v1alpha1
```
can we remove this as well. this comment also applies for L437-438

miguelbernadi replied:
Golint still shows failures for these in my local machine. There are non-autogenerated files called `defaults.go` with golint violations because they invoke autogenerated function names. Example:
```
plugin/pkg/admission/eventratelimit/apis/eventratelimit/v1alpha1/defaults.go:25:6: don't use underscores in Go names; func SetDefaults_Configuration should be SetDefaultsConfiguration
```
I can try and see if the CI will ignore these violations.

yue9944882 replied:
> There are non-autogenerated files called defaults.go with golint violations because they invoke autogenerated function names.

oh. apologises. lets keep it if it's generated code

miguelbernadi replied:
Some of the complaints can be fixed, but the ones on using underscores require changing how code is generated. Failure results:
```
Errors from golint:
plugin/pkg/admission/eventratelimit/apis/eventratelimit/v1alpha1/defaults.go:25:1: exported function SetDefaults_Configuration should have comment or be unexported
plugin/pkg/admission/eventratelimit/apis/eventratelimit/v1alpha1/defaults.go:25:6: don't use underscores in Go names; func SetDefaults_Configuration should be SetDefaultsConfiguration
plugin/pkg/admission/resourcequota/apis/resourcequota/v1alpha1/defaults.go:25:1: exported function SetDefaults_Configuration should have comment or be unexported
plugin/pkg/admission/resourcequota/apis/resourcequota/v1alpha1/defaults.go:25:6: don't use underscores in Go names; func SetDefaults_Configuration should be SetDefaultsConfiguration
plugin/pkg/admission/resourcequota/apis/resourcequota/v1beta1/defaults.go:25:1: exported function SetDefaults_Configuration should have comment or be unexported
plugin/pkg/admission/resourcequota/apis/resourcequota/v1beta1/defaults.go:25:6: don't use underscores in Go names; func SetDefaults_Configuration should be SetDefaultsConfiguration
```

miguelbernadi replied:
I removed the commit that verified this breaks in CI.

yue9944882 reviewed (approved):
/lgtm

thx for fixing these! 🎉

/assign @deads2k 
for approval

miguelbernadi commented:
@deads2k Any input on the PR?

miguelbernadi commented:
@deads2k ping

miguelbernadi commented:
/test pull-kubernetes-verify

miguelbernadi commented:
/test pull-kubernetes-integration

yue9944882 commented:
/lgtm

miguelbernadi commented:
Any news on this front? I'm ready to rebase or complete it so it can be merged.

mourya007 commented on plugin/pkg/admission/noderestriction/admission.go:
```diff
@@ -161,7 +163,7 @@ func (c *nodePlugin) Admit(a admission.Attributes) error {
 }
 
-func (c *nodePlugin) admitPod(nodeName string, a admission.Attributes) error {
+func (p *Plugin) admitPod(nodeName string, a admission.Attributes) error {
```
Can we add golint comment  for the admitpod() also?

miguelbernadi replied:
I can add it, but private functions are usually not required to have one.

mourya007 replied:
yeah, right but I guess doing same on the private function under the same file is not a big issue. What do you think?

mourya007 commented on plugin/pkg/admission/noderestriction/admission.go:
```diff
@@ -222,7 +224,7 @@ func (c *nodePlugin) admitPod(nodeName string, a admission.Attributes) error {
 }
 
-func (c *nodePlugin) admitPodStatus(nodeName string, a admission.Attributes) error {
+func (p *Plugin) admitPodStatus(nodeName string, a admission.Attributes) error {
```
Ditto

mourya007 commented on plugin/pkg/admission/noderestriction/admission.go:
```diff
@@ -241,7 +243,7 @@ func (c *nodePlugin) admitPodStatus(nodeName string, a admission.Attributes) err
 }
 
-func (c *nodePlugin) admitPodEviction(nodeName string, a admission.Attributes) error {
+func (p *Plugin) admitPodEviction(nodeName string, a admission.Attributes) error {
```
here too.

mourya007 commented:
/retest

miguelbernadi commented:
I'll get to teh rebase and addressing the new comments later today.

miguelbernadi commented:
@mourya007 Just finished the rebase. Will be addressing your comments now. Bear in mind those methods are unexported, so the comment is usually not required (and golint does not enforce it either).

mourya007 commented on plugin/pkg/admission/noderestriction/admission.go:
```diff
@@ -442,7 +449,7 @@ func (c *nodePlugin) getForbiddenCreateLabels(modifiedLabels sets.String) sets.S
 }
 
 // getForbiddenLabels returns the set of labels that may not be set by the node on update.
```
getForbiddenUpdateLabels ?

miguelbernadi replied:
I'll update this comment. Can you confirm the descriptions are correct?

miguelbernadi commented:
/test pull-kubernetes-e2e-kops-aws

miguelbernadi commented:
/test pull-kubernetes-e2e-kops-aws

miguelbernadi commented:
/test pull-kubernetes-e2e-kops-aws

miguelbernadi commented:
/test pull-kubernetes-e2e-kops-aws

miguelbernadi commented:
/test pull-kubernetes-e2e-kops-aws

dims commented:
/assign @liggitt 
/milestone v1.15

/approve

(approve for the hack/ changes)

liggitt commented:
/approve

/hold
please squash before merge

miguelbernadi commented:
Commits have been squashed. Thanks to everyone that helped!

liggitt commented on plugin/pkg/admission/limitranger/interfaces.go:
```diff
@@ -14,14 +14,15 @@ See the License for the specific language governing permissions and
 */
 
-package limitranger
+package limitrangeradmission
```
please revert this change

miguelbernadi replied:
The rename was needed to prevent failure of this golint rule:
```
Errors from golint:
plugin/pkg/admission/limitranger/interfaces.go:26:6: type name will be used as limitranger.LimitRangerActions by other packages, and that stutters; consider calling this Actions
```
I'm open to suggestions on fixing it. The original approach keeps the same use in all clients and just renames it internally.

liggitt replied:
I'm not in favor of changing our package and type names to satisfy that lint recommendation

miguelbernadi replied:
One option would be to rename the type, like the linter suggests, from LimitRangerActions to Actions. Another option is to keep this one blacklisted and unchecked until a future time so we can all agree on a valid solution. Which would you prefer?

yue9944882 replied:
i'm fine w/ both. it's just the smaller change looks more happy.

miguelbernadi replied:
I've removed the renaming with its effects on all files and restored the blacklist on the limitranger package. I've kept all other linting fixes so it's a smaller effort once there's an agreement on what to do for this particular issue.

liggitt replied:
> I've removed the renaming with its effects on all files and restored the blacklist on the limitranger package. I've kept all other linting fixes so it's a smaller effort once there's an agreement on what to do for this particular issue.

sounds good, thanks

liggitt commented:
one comment on the package rename, lgtm otherwise

liggitt commented:
/lgtm
/approve

yue9944882 commented:
/retest

miguelbernadi commented:
What's left to merge this PR? Anything I can help with?

yue9944882 commented:
/hold cancel

as @liggitt's comments are all addressed

miguelbernadi commented:
/retest

miguelbernadi commented:
Seems we need an `/ok-to-test` for tests to run again. @yue9944882

yue9944882 commented:
/retest
/ok-to-test

liggitt commented:
if it makes the milestone, that's fine, but is not required
